File: Emulator/src/main/core/game/interaction/DestinationFlag.java
package core.game.interaction;

import core.game.global.action.DoorActionHandler;
import core.game.node.Node;
import core.game.node.entity.Entity;
import core.game.node.scenery.Scenery;
import core.game.world.map.Direction;
import core.game.world.map.Location;
import core.game.world.map.RegionManager;

/**
 * The type Destination flag.
 */
public class DestinationFlag {

    /**
     * The constant LOCATION.
     */
    public static final DestinationFlag LOCATION = new DestinationFlag();

    /**
     * The constant ENTITY.
     */
    public static final DestinationFlag ENTITY = new DestinationFlag() {

        @Override
        public Location getDestination(Entity mover, Node n) {
            Location l = getClosestTo(mover, n, n.getLocation().transform(0, -1, 0));
            if (mover.size() > 1) {
                if (l.getX() < n.getLocation().getX()) {
                    l = l.transform(-(mover.size() - 1), 0, 0);
                }
                if (l.getY() < n.getLocation().getY()) {
                    l = l.transform(0, -(mover.size() - 1), 0);
                }
            }
            return l;
        }
    };

    /**
     * The constant FOLLOW_ENTITY.
     */
    public static final DestinationFlag FOLLOW_ENTITY = new DestinationFlag() {

        @Override
        public Location getDestination(Entity mover, Node n) {
            Direction dir = n.getDirection();
            Location l = n.getLocation().transform(-dir.getStepX(), -dir.getStepY(), 0);
            if (!checkTraversal(l, dir)) {
                l = getClosestTo(mover, n, l);
            }
            if (mover.size() > 1) {
                if (l.getX() < n.getLocation().getX()) {
                    l = l.transform(-(mover.size() - 1), 0, 0);
                }
                if (l.getY() < n.getLocation().getY()) {
                    l = l.transform(0, -(mover.size() - 1), 0);
                }
            }
            return l;
        }
    };

    /**
     * The constant COMBAT.
     */
    public static final DestinationFlag COMBAT = new DestinationFlag() {

        @Override
        public Location getDestination(Entity mover, Node n) {
            return null; // Placeholder for combat-related movement
        }
    };

    /**
     * The constant ITEM.
     */
    public static final DestinationFlag ITEM = new DestinationFlag() {

        @Override
        public Location getDestination(Entity mover, Node n) {
            if (!RegionManager.isTeleportPermitted(n.getLocation())) {
                return getClosestTo(mover, n, n.getLocation().transform(1, 0, 0));
            }
            return n.getLocation();
        }
    };

    /**
     * The constant OBJECT.
     */
    public static final DestinationFlag OBJECT = new DestinationFlag() {

        @Override
        public Location getDestination(Entity mover, Node n) {
            Scenery scenery = (Scenery) n;
            if (scenery.getType() < 4 || scenery.getType() == 9) {
                return DoorActionHandler.getDestination(mover, scenery);
            }
            if (scenery.getType() == 4 || scenery.getType() == 5) {
                // Wall or decoration
                return scenery.getLocation();
            }
            int sizeX = scenery.getDefinition().sizeX;
            int sizeY = scenery.getDefinition().sizeY;
            if (scenery.getRotation() % 2 != 0) {
                int switcher = sizeX;
                sizeX = sizeY;
                sizeY = switcher;
            }
            Direction dir = Direction.forWalkFlag(scenery.getDefinition().walkingFlag, scenery.getRotation());
            if (dir != null) {
                return getDestination(mover, scenery, sizeX, sizeY, dir, 3);
            }

            return getDestination(mover, scenery, sizeX, sizeY, Direction.getLogicalDirection(scenery.getLocation(), mover.getLocation()), 0);
        }

        private Location getDestination(Entity mover, Scenery scenery, int sizeX, int sizeY, Direction dir, int count) {
            Location closest = null;
            double distance = 9999.9;
            Location loc = scenery.getLocation();
            for (int i = count; i < 4; i++) {
                if (dir.toInteger() % 2 != 0) {
                    int x = dir.getStepX();
                    if (x > 0) {
                        x *= sizeX;
                    }
                    for (int y = 0; y < sizeY; y++) {
                        Location l = loc.transform(x, y, 0);
                        if (checkTraversal(l, dir)) {
                            double dist = mover.getLocation().getDistance(l);
                            if (dist < distance) {
                                distance = dist;
                                closest = l;
                            }
                        }
                    }
                } else {
                    int y = dir.getStepY();
                    if (y > 0) {
                        y *= sizeY;
                    }
                    for (int x = 0; x < sizeX; x++) {
                        Location l = loc.transform(x, y, 0);
                        if (checkTraversal(l, dir)) {
                            double dist = mover.getLocation().getDistance(l);
                            if (dist < distance) {
                                distance = dist;
                                closest = l;
                            }
                        }
                    }
                }
                dir = Direction.get((dir.toInteger() + 1) % 4);
            }
            return closest;
        }

        @Override
        public boolean checkTraversal(Location l, Direction dir) {
            return RegionManager.isTeleportPermitted(l) && dir.canMove(l);
        }
    };

    /**
     * Instantiates a new Destination flag.
     */
    public DestinationFlag() {
        // Empty
    }

    /**
     * Gets destination.
     *
     * @param mover the mover
     * @param node  the node
     * @return the destination
     */
    public Location getDestination(Entity mover, Node node) {
        return node.getLocation();
    }

    /**
     * Check traversal boolean.
     *
     * @param l   the l
     * @param dir the dir
     * @return the boolean
     */
    public boolean checkTraversal(Location l, Direction dir) {
        return Direction.get((dir.toInteger() + 2) % 4).canMove(l);
    }

    /**
     * Gets closest to.
     *
     * @param mover      the mover
     * @param node       the node
     * @param suggestion the suggestion
     * @return the closest to
     */
    public Location getClosestTo(Entity mover, Node node, Location suggestion) {
        Location nl = node.getLocation();
        int diffX = suggestion.getX() - nl.getX();
        int diffY = suggestion.getY() - nl.getY();
        Direction moveDir = Direction.NORTH;
        if (diffX < 0) {
            moveDir = Direction.EAST;
        } else if (diffX >= node.size()) {
            moveDir = Direction.WEST;
        } else if (diffY >= node.size()) {
            moveDir = Direction.SOUTH;
        }
        double distance = 9999.9;
        Location destination = suggestion;
        for (int c = 0; c < 4; c++) {
            for (int i = 0; i < node.size() + 1; i++) {
                for (int j = 0; j < (i == 0 ? 1 : 2); j++) {
                    Direction current = Direction.get((moveDir.toInteger() + (j == 1 ? 3 : 1)) % 4);
                    Location loc = suggestion.transform(current.getStepX() * i, current.getStepY() * i, 0);
                    if (moveDir.toInteger() % 2 == 0) {
                        if (loc.getX() < nl.getX() || loc.getX() > nl.getX() + node.size() - 1) {
                            continue;
                        }
                    } else {
                        if (loc.getY() < nl.getY() || loc.getY() > nl.getY() + node.size() - 1) {
                            continue;
                        }
                    }
                    if (checkTraversal(loc, moveDir)) {
                        double dist = mover.getLocation().getDistance(loc);
                        if (dist < distance) {
                            distance = dist;
                            destination = loc;
                        }
                    }
                }
            }
            moveDir = Direction.get((moveDir.toInteger() + 1) % 4);
            int offsetX = Math.abs(moveDir.getStepY() * (node.size() >> 1));
            int offsetY = Math.abs(moveDir.getStepX() * (node.size() >> 1));
            if (moveDir.toInteger() < 2) {
                suggestion = node.getLocation().transform(-moveDir.getStepX() + offsetX, -moveDir.getStepY() + offsetY, 0);
            } else {
                suggestion = node.getLocation().transform(-moveDir.getStepX() * node.size() + offsetX, -moveDir.getStepY() * node.size() + offsetY, 0);
            }
        }
        return destination;
    }
}

File: Server/src/main/core/game/interaction/DestinationFlag.kt
package core.game.interaction

import core.game.global.action.DoorActionHandler
import core.game.node.Node
import core.game.node.entity.Entity
import core.game.node.scenery.Scenery
import core.game.world.map.Direction
import core.game.world.map.Location
import core.game.world.map.RegionManager
import kotlin.math.abs

/**
 * Handles pathfinding destination logic for different interaction types.
 */
open class DestinationFlag {

    /**
     * Gets the default destination location for a mover interacting with a node.
     *
     * @param mover The moving entity (player or NPC).
     * @param node The target node.
     * @return The destination location.
     */
    open fun getDestination(mover: Entity, node: Node): Location {
        return node.location
    }

    /**
     * Checks whether traversal is possible from a given location in the given direction.
     *
     * @param l The location to check.
     * @param dir The direction of intended movement.
     * @return True if traversal is permitted, false otherwise.
     */
    open fun checkTraversal(l: Location, dir: Direction): Boolean {
        return Direction.get((dir.toInteger() + 2) % 4).canMove(l)
    }

    /**
     * Finds the closest traversable tile to the given suggestion near the node.
     *
     * @param mover The moving entity.
     * @param node The node the mover is approaching.
     * @param suggestion The preferred destination to test from.
     * @return The closest reachable tile near the node.
     */
    fun getClosestTo(mover: Entity, node: Node, suggestion: Location): Location {
        val nl = node.location
        val diffX = suggestion.x - nl.x
        val diffY = suggestion.y - nl.y
        var moveDir = Direction.NORTH
        if (diffX < 0) {
            moveDir = Direction.EAST
        } else if (diffX >= node.size()) {
            moveDir = Direction.WEST
        } else if (diffY >= node.size()) {
            moveDir = Direction.SOUTH
        }
        var distance = 9999.9
        var destination = suggestion
        var suggest = suggestion
        for (c in 0 until 4) {
            for (i in 0..node.size()) {
                for (j in 0 until if (i == 0) 1 else 2) {
                    val current = Direction.get((moveDir.toInteger() + if (j == 1) 3 else 1) % 4)
                    val loc = suggest.transform(current.stepX * i, current.stepY * i, 0)
                    if (moveDir.toInteger() % 2 == 0) {
                        if (loc.x < nl.x || loc.x > nl.x + node.size() - 1) continue
                    } else {
                        if (loc.y < nl.y || loc.y > nl.y + node.size() - 1) continue
                    }
                    if (checkTraversal(loc, moveDir)) {
                        val dist = mover.location.getDistance(loc)
                        if (dist < distance) {
                            distance = dist
                            destination = loc
                        }
                    }
                }
            }
            moveDir = Direction.get((moveDir.toInteger() + 1) % 4)
            val offsetX = abs(moveDir.stepY * (node.size() shr 1))
            val offsetY = abs(moveDir.stepX * (node.size() shr 1))
            suggest = if (moveDir.toInteger() < 2) {
                node.location.transform(-moveDir.stepX + offsetX, -moveDir.stepY + offsetY, 0)
            } else {
                node.location.transform(
                    -moveDir.stepX * node.size() + offsetX, -moveDir.stepY * node.size() + offsetY, 0
                )
            }
        }
        return destination
    }

    companion object {
        /**
         * Default destination: simply returns the node's location.
         */
        @JvmField
        val LOCATION = DestinationFlag()

        /**
         * Destination when moving to interact with an entity (e.g. player or NPC).
         */
        @JvmField
        val ENTITY = object : DestinationFlag() {
            override fun getDestination(mover: Entity, node: Node): Location {
                var l = getClosestTo(mover, node, node.location.transform(0, -1, 0))
                if (mover.size() > 1) {
                    if (l.x < node.location.x) {
                        l = l.transform(-(mover.size() - 1), 0, 0)
                    }
                    if (l.y < node.location.y) {
                        l = l.transform(0, -(mover.size() - 1), 0)
                    }
                }
                return l
            }
        }

        /**
         * Destination for following an entity, adjusting position based on their facing direction.
         */
        @JvmField
        val FOLLOW_ENTITY = object : DestinationFlag() {
            override fun getDestination(mover: Entity, node: Node): Location {
                val dir = node.direction
                var l = node.location.transform(-dir.stepX, -dir.stepY, 0)
                if (!checkTraversal(l, dir)) {
                    l = getClosestTo(mover, node, l)
                }
                if (mover.size() > 1) {
                    if (l.x < node.location.x) {
                        l = l.transform(-(mover.size() - 1), 0, 0)
                    }
                    if (l.y < node.location.y) {
                        l = l.transform(0, -(mover.size() - 1), 0)
                    }
                }
                return l
            }
        }

        /**
         * Placeholder for combat-related movement destinations.
         */
        @JvmField
        val COMBAT = object : DestinationFlag() {
            override fun getDestination(mover: Entity, node: Node): Location {
                return node.location
            }
        }

        /**
         * Destination when walking toward a ground item.
         */
        @JvmField
        val ITEM = object : DestinationFlag() {
            override fun getDestination(mover: Entity, node: Node): Location {
                return if (!RegionManager.isTeleportPermitted(node.location)) {
                    getClosestTo(mover, node, node.location.transform(1, 0, 0))
                } else {
                    node.location
                }
            }
        }

        /**
         * Destination when interacting with a game object (scenery).
         */
        @JvmField
        val OBJECT = object : DestinationFlag() {
            override fun getDestination(mover: Entity, node: Node): Location {
                val scenery = node as Scenery
                return when (scenery.type) {
                    in 0..3, 9 -> DoorActionHandler.getDestination(mover, scenery)
                    4, 5 -> scenery.location
                    else -> {
                        var sizeX = scenery.definition.sizeX
                        var sizeY = scenery.definition.sizeY
                        if (scenery.rotation % 2 != 0) {
                            val tmp = sizeX
                            sizeX = sizeY
                            sizeY = tmp
                        }
                        val dir = Direction.forWalkFlag(scenery.definition.blockFlag, scenery.rotation)
                        val destination = if (dir != null) {
                            getDestination(mover, scenery, sizeX, sizeY, dir, 3)
                        } else {
                            getDestination(
                                mover,
                                scenery,
                                sizeX,
                                sizeY,
                                Direction.getLogicalDirection(scenery.location, mover.location),
                                0
                            )
                        }
                        return destination ?: scenery.location
                    }
                }
            }

            /**
             * Finds the closest traversable tile around an object, respecting size and direction.
             *
             * @param mover The moving entity.
             * @param scenery The object being interacted with.
             * @param sizeX Width of the object.
             * @param sizeY Height of the object.
             * @param startDir Starting direction to check.
             * @param count Rotation offset (used when direction is known).
             * @return The closest reachable tile near the object, or null if none found.
             */
            private fun getDestination(mover: Entity, scenery: Scenery, sizeX: Int, sizeY: Int, startDir: Direction, count: Int): Location? {
                var closest: Location? = null
                var distance = 9999.9
                val loc = scenery.location
                var dir = startDir
                for (i in count until 4) {
                    if (dir.toInteger() % 2 != 0) {
                        val x = if (dir.stepX > 0) dir.stepX * sizeX else dir.stepX
                        for (y in 0 until sizeY) {
                            val l = loc.transform(x, y, 0)
                            if (checkTraversal(l, dir)) {
                                val dist = mover.location.getDistance(l)
                                if (dist < distance) {
                                    distance = dist
                                    closest = l
                                }
                            }
                        }
                    } else {
                        val y = if (dir.stepY > 0) dir.stepY * sizeY else dir.stepY
                        for (x in 0 until sizeX) {
                            val l = loc.transform(x, y, 0)
                            if (checkTraversal(l, dir)) {
                                val dist = mover.location.getDistance(l)
                                if (dist < distance) {
                                    distance = dist
                                    closest = l
                                }
                            }
                        }
                    }
                    dir = Direction.get((dir.toInteger() + 1) % 4)
                }
                return closest
            }

            /**
             * Checks if the tile in the given direction is traversable and not blocked.
             */
            override fun checkTraversal(l: Location, dir: Direction): Boolean {
                return RegionManager.isTeleportPermitted(l) && dir.canMove(l)
            }
        }
    }
}